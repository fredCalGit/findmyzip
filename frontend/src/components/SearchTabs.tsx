import React, {
  Dispatch,
  SetStateAction,
  SyntheticEvent,
  useState,
} from "react";
import {
  Box,
  Card,
  CardContent,
  CardHeader,
  Skeleton,
  Tab,
  Tabs,
  Typography,
} from "@mui/material";

export interface SearchTabsProps {
  data: {
    location: {
      country: string;
      countryAbbreviation: string;
      postCode: string;
      places: {
        placeName: string;
        state: string;
        stateAbbreviation: string;
        latitude: string;
        longitude: string;
      }[];
    };
  };
  cachedVariables?: { country: string; postCode: string }[] | null;
  setQueryVariables: Dispatch<
    SetStateAction<{ country: string; postCode: string } | null>
  >;
}

export const SearchTabs = ({
  data,
  cachedVariables = null,
  setQueryVariables,
}: SearchTabsProps) => {
  const [value, setValue] = useState(0);

  const handleChange = (event: SyntheticEvent, newValue: number) => {
    setValue(newValue);
  };
  return (
    <Box sx={{ width: "100%" }}>
      <Box sx={{ borderBottom: 1, borderColor: "divider" }}>
        {cachedVariables && (
          <Tabs
            value={cachedVariables.length > 0 ? cachedVariables.length - 1 : 0}
            onChange={handleChange}
            aria-label='basic tabs example'
          >
            {data ? (
              cachedVariables &&
              cachedVariables.map((item, index) => {
                return (
                  <Tab
                    key={`${item.country}-${item.postCode}-${index}`}
                    label={`${item.country} - ${item.postCode}`}
                    onClick={() => {
                      setQueryVariables(cachedVariables[index]);
                    }}
                    {...a11yProps(index)}
                  />
                );
              })
            ) : (
              <Skeleton />
            )}
          </Tabs>
        )}
      </Box>

      {data ? (
        cachedVariables &&
        cachedVariables.map((item, index) => {
          return (
            <TabPanel
              value={value}
              index={index}
              key={`${item.postCode}-${item.country}-${index}`}
            >
              {data.location ? (
                <>
                  <Typography variant='h5'>
                    Country: {data.location.country} -{" "}
                    {data.location.countryAbbreviation}
                  </Typography>
                  <Typography variant='h5'>
                    ZipCode: {data.location.postCode}
                  </Typography>
                </>
              ) : (
                <Typography color='error' variant='h5'>
                  Zip Code not found!
                </Typography>
              )}

              {data?.location?.places ? (
                data.location.places.map((item, index) => {
                  return (
                    <Card key={`${item.longitude}-${item.latitude}`}>
                      <CardHeader
                        title={`${item.placeName} - ${item.stateAbbreviation}`}
                      />
                      <CardContent>
                        <Typography>State: {item.state}</Typography>
                        <Typography>Latitude: {item.latitude}</Typography>
                        <Typography>Longitude: {item.longitude}</Typography>
                      </CardContent>
                    </Card>
                  );
                })
              ) : (
                <Typography color='error'>Enter Valid Zip Code!</Typography>
              )}
            </TabPanel>
          );
        })
      ) : (
        <Skeleton />
      )}
    </Box>
  );
};

interface TabPanelProps {
  children?: React.ReactNode;
  index: number;
  value: number;
}
function TabPanel(props: TabPanelProps) {
  const { children, value, index, ...other } = props;

  return (
    <div
      role='tabpanel'
      hidden={value !== index}
      id={`simple-tabpanel-${index}`}
      aria-labelledby={`simple-tab-${index}`}
      {...other}
    >
      {value === index && <Box sx={{ p: 3 }}>{children}</Box>}
    </div>
  );
}

function a11yProps(index: number) {
  return {
    id: `simple-tab-${index}`,
    "aria-controls": `simple-tabpanel-${index}`,
  };
}
