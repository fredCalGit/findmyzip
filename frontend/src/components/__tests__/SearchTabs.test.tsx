import React from "react";
import { render } from "@testing-library/react";
import { SearchTabs } from "../SearchTabs";

const mockData = {
  location: {
    country: "United States",
    countryAbbreviation: "US",
    postCode: "90211",
    places: [
      {
        placeName: "Beverly Hills",
        state: "California",
        stateAbbreviation: "CA",
        latitude: "34.0652",
        longitude: "-118.383",
      },
    ],
  },
};

const mockVariables = {
  country: "US",
  postCode: "90211",
};
const mockCachedVariables = [mockVariables];

test("renders the SearchTab", () => {
  const setState = jest.fn();

  const { container } = render(
    <SearchTabs
      data={mockData}
      cachedVariables={mockCachedVariables}
      setQueryVariables={setState}
    />
  );
  expect(container).toMatchSnapshot();
});
