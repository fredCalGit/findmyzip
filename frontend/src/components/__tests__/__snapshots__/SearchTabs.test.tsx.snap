// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the SearchTab 1`] = `
<div>
  <div
    class="MuiBox-root css-8atqhb"
  >
    <div
      class="MuiBox-root css-1gsv261"
    >
      <div
        class="MuiTabs-root css-1ujnqem-MuiTabs-root"
      >
        <div
          class="MuiTabs-scroller MuiTabs-fixed css-jpln7h-MuiTabs-scroller"
          style="overflow: hidden; margin-bottom: 0px;"
        >
          <div
            aria-label="basic tabs example"
            class="MuiTabs-flexContainer css-heg063-MuiTabs-flexContainer"
            role="tablist"
          >
            <button
              aria-controls="simple-tabpanel-0"
              aria-selected="true"
              class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary Mui-selected css-1h9z7r5-MuiButtonBase-root-MuiTab-root"
              id="simple-tab-0"
              role="tab"
              tabindex="0"
              type="button"
            >
              US - 90211
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
          <span
            class="MuiTabs-indicator css-1aquho2-MuiTabs-indicator"
            style="left: 0px; width: 0px;"
          />
        </div>
      </div>
    </div>
    <div
      aria-labelledby="simple-tab-0"
      id="simple-tabpanel-0"
      role="tabpanel"
    >
      <div
        class="MuiBox-root css-19kzrtu"
      >
        <h5
          class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
        >
          Country: 
          United States
           -
           
          US
        </h5>
        <h5
          class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
        >
          ZipCode: 
          90211
        </h5>
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
        >
          <div
            class="MuiCardHeader-root css-185gdzj-MuiCardHeader-root"
          >
            <div
              class="MuiCardHeader-content css-1qbkelo-MuiCardHeader-content"
            >
              <span
                class="MuiTypography-root MuiTypography-h5 MuiCardHeader-title css-1qvr50w-MuiTypography-root"
              >
                Beverly Hills - CA
              </span>
            </div>
          </div>
          <div
            class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            >
              State: 
              California
            </p>
            <p
              class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            >
              Latitude: 
              34.0652
            </p>
            <p
              class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            >
              Longitude: 
              -118.383
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
